server/etcdserver/api/etcdhttp/health.go

```diff
@@ -194,3 +201,183 @@ func checkAPI(lg *zap.Logger, srv ServerHealth, serializable bool) Health {
 	lg.Debug("serving /health true")
 	return h
 }
+
+type HealthCheck func(ctx context.Context) error
+
+type CheckRegistry struct {
+	path string
+
+	checks map[string]HealthCheck
+	// the list of checkNames are needed to ensure the order of checks.
+	checkNames []string
+}
+
+func installLivezEndpoints(lg *zap.Logger, mux *http.ServeMux, server ServerHealth) {
+	reg := CheckRegistry{path: "/livez", checks: make(map[string]HealthCheck)}
+	reg.register("serializable_read", serializableReadCheck(server))
+	reg.installHttpEndpoints(lg, mux)
+}
+
+func installReayzEndpoints(lg *zap.Logger, mux *http.ServeMux, server ServerHealth) {
+	reg := CheckRegistry{path: "/readyz", checks: make(map[string]HealthCheck)}
+	reg.register("data_corruption", activeAlarmCheck(server, etcdserverpb.AlarmType_CORRUPT))
+	reg.register("serializable_read", serializableReadCheck(server))
+	reg.installHttpEndpoints(lg, mux)
+}
+
+func (reg *CheckRegistry) register(name string, check HealthCheck) {
+	reg.checkNames = append(reg.checkNames, name)
+	reg.checks[name] = check
+}
+
+func (reg *CheckRegistry) installHttpEndpoints(lg *zap.Logger, mux *http.ServeMux) {
+	lg.Info(fmt.Sprintf("Installing %s http endpoint.", reg.path))
+	mux.Handle(reg.path, reg.handleRootHealth(lg))
+	for checkName, check := range reg.checks {
+		reg.installCheckEndpoint(checkName, check, mux)
+	}
+}
+
+func (reg *CheckRegistry) runHealthChecks(ctx context.Context, checkNames []string) (Health, error) {
+	h := Health{Health: "true"}
+	failedChecks := []string{}
+	var individualCheckOutput bytes.Buffer
+	for _, checkName := range checkNames {
+		check, found := reg.checks[checkName]
+		if !found {
+			return h, fmt.Errorf("Health check: %s not registered", checkName)
+		}
+		if err := check(ctx); err != nil {
+			fmt.Fprintf(&individualCheckOutput, "[-]%s failed: %v\n", checkName, err)
+			failedChecks = append(failedChecks, checkName)
+		} else {
+			fmt.Fprintf(&individualCheckOutput, "[+]%s ok\n", checkName)
+		}
+	}
+	h.Reason = individualCheckOutput.String()
+	// always be verbose on failure
+	if len(failedChecks) > 0 {
+		h.Health = "false"
+	}
+	return h, nil
+}
+
+// handleRootHealth returns an http.HandlerFunc that serves the provided checks.
+func (reg *CheckRegistry) handleRootHealth(lg *zap.Logger) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// extracts the health check names to be excludeList from the query param
+		excluded := getQuerySet(r, "exclude")
+		// extracts the health check names to be allowList from the query param
+		included := getQuerySet(r, "allowlist")
+
+		filteredCheckNames, err := filterCheckList(reg.checkNames, included, excluded)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+		h, _ := reg.runHealthChecks(r.Context(), filteredCheckNames)
```

Comment:
Since the Line 278 doesn't check the error, the `h.Health` can be true I think.
I think it's unexpected.